File: src/main/java/hzt/aoc/day12/Part1RainRisk.java
package hzt.aoc.day12;

import java.awt.*;
import java.util.List;

public class Part1RainRisk extends Day12Challenge {

    public Part1RainRisk() {
        super("part 1",
                "What is the Manhattan distance between that location and the ship's starting position? See ChallengeDAy12.md");
    }

    @Override
    protected Object solve(List<String> inputList) {
        Point position = new Point(0, 0);
        Point orientation = new Point(1, 0);
        for (String line : inputList) {
            char instruction = line.charAt(0);
            int amount = Integer.parseInt(line.substring(1));
            orientation = rotateShip(orientation, instruction, amount);
            translatePointInWindDirection(position, instruction, amount);
            if (instruction == MOVE_FORWARD) position.translate(orientation.x * amount, orientation.y * amount);
        }
        return getMessage(Math.abs(position.x) + Math.abs(position.y));
    }

    private Point rotateShip(Point orientation, char instruction, int angle) {
        if ((instruction == TURN_RIGHT || instruction == TURN_LEFT) && angle % NINETY_DEGREES == 0) {
            int dir = instruction == TURN_RIGHT ? 1 : -1;
            int counter = 0;
            while (angle / NINETY_DEGREES != counter) {
                if (orientation.equals(new Point(1, 0))) orientation = new Point(0, -dir);
                else if (orientation.equals(new Point(0, 1))) orientation = new Point(dir, 0);
                else if (orientation.equals(new Point(-1, 0))) orientation = new Point(0, dir);
                else if (orientation.equals(new Point(0, -1))) orientation = new Point(-dir, 0);
                counter++;
            }
        }
        return orientation;
    }

    @Override
    String getMessage(int global) {
        return String.format("%d", global);
    }
}

File: src/main/java/hzt/aoc/day12/Part1RainRisk.kt
package hzt.aoc.day12

import java.awt.Point
import kotlin.math.abs

class Part1RainRisk : Day12Challenge(
    "part 1",
    "What is the Manhattan distance between that location and the ship's starting position See ChallengeDAy12.md"
) {
    override fun solve(inputList: List<String>): String {
        val position = Point(0, 0)
        var orientation = Point(1, 0)
        for (line in inputList) {
            val instruction = line[0]
            val amount = line.substring(1).toInt()
            orientation = rotateShip(orientation, instruction, amount)
            translatePointInWindDirection(position, instruction, amount)
            if (instruction == MOVE_FORWARD) position.translate(
                orientation.x * amount,
                orientation.y * amount
            )
        }
        return getMessage(abs(position.x) + abs(position.y))
    }

    private fun rotateShip(orientation: Point, instruction: Char, angle: Int): Point {
        var newOrientation = orientation
        if ((instruction == TURN_RIGHT || instruction == TURN_LEFT) && angle % NINETY_DEGREES == 0) {
            val dir = if (instruction == TURN_RIGHT) 1 else -1
            var counter = 0
            while (angle / NINETY_DEGREES != counter) {
                when (newOrientation) {
                    Point(1, 0) -> newOrientation = Point(0, -dir)
                    Point(0, 1) -> newOrientation = Point(dir, 0)
                    Point(-1, 0) -> newOrientation = Point(0, dir)
                    Point(0, -1) -> newOrientation = Point(-dir, 0)
                }
                counter++
            }
        }
        return newOrientation
    }

    override fun getMessage(value: Int): String = String.format("%d", value)
}
